query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+/// ! type tags out of the store at runtime.
+/// ! `wheres` is a list of fragments that can be joined by `:and`.
+
+extern crate mentat_core;
+extern crate mentat_query;
+
+use ::std::collections::BTreeMap;
+use ::std::collections::btree_map::Entry;
+
+use self::mentat_core::{
+    Attribute,
+    Entid,
+    Schema,
+    TypedValue,
+    ValueType,
+};
+
+use self::mentat_query::{
+    IdentOrEntid,
+    NamespacedKeyword,
+    NonIntegerConstant,
+    Pattern,
+    PatternNonValuePlace,
+    PatternValuePlace,
+    PlainSymbol,
+    SrcVar,
+    Variable,
+};
+
+/// This enum models the fixed set of default tables we have -- two
+/// tables and two views.
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum DatomsTable {
+    Datoms,             // The non-fulltext datoms table.
+    FulltextValues,     // The virtual table mapping IDs to strings.
+    FulltextDatoms,     // The fulltext-datoms view.
+    AllDatoms,          // Fulltext and non-fulltext datoms.
+}
+
+impl DatomsTable {
+    fn name(&self) -> &'static str {
+        match *self {
+            DatomsTable::Datoms => "datoms",
+            DatomsTable::FulltextValues => "fulltext_values",
+            DatomsTable::FulltextDatoms => "fulltext_datoms",
+            DatomsTable::AllDatoms => "all_datoms",
+        }
+    }
+}
+
+/// One of the named columns of our tables.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum DatomsColumn {
+    Entity,
+    Attribute,
+    Value,
+    Tx,
+    ValueTypeTag,
+}
+
+/// A specific instance of a table within a query. E.g., "datoms123".
+pub type TableAlias = String;
+
+/// The association between a table and its alias. E.g., AllDatoms, "all_datoms123".
+#[derive(PartialEq, Eq, Debug)]
+pub struct SourceAlias(DatomsTable, TableAlias);
+
+/// A particular column of a particular aliased table. E.g., "datoms123", Attribute.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct QualifiedAlias(TableAlias, DatomsColumn);
+
+impl QualifiedAlias {
+    fn for_type_tag(&self) -> QualifiedAlias {
+        QualifiedAlias(self.0.clone(), DatomsColumn::ValueTypeTag)
+    }
+
+    #[inline]
+    fn is_value(&self) -> bool {
+        self.1 == DatomsColumn::Value
+    }
+}
+
+/// A thing that's capable of aliasing a table name for us.
+/// This exists so that we can obtain predictable names in tests.
+pub type TableAliaser = Box<FnMut(DatomsTable) -> TableAlias>;
+
+pub fn default_table_aliaser() -> TableAliaser {
+    let mut i = -1;
+    Box::new(move |table| {
+        i += 1;
+        format!("{}{}", table.name(), i)
+    })
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ColumnConstraint {
+    EqualsEntity(QualifiedAlias, Entid),
+    EqualsValue(QualifiedAlias, TypedValue),
+    EqualsColumn(QualifiedAlias, QualifiedAlias),
+}
+
+pub struct ConjoiningClauses {
+    is_known_empty: bool,
+    aliaser: TableAliaser,
+    from: Vec<SourceAlias>,
+    wheres: Vec<ColumnConstraint>,
+    bindings: BTreeMap<Variable, Vec<QualifiedAlias>>,
```

Comment:
I'm not at all comfortable cloning! At some point we do indeed need to have `Rc<Variable>` instead.